docs/tutorial-part2.rst

```diff
@@ -0,0 +1,550 @@
+
+.. code-block:: python
+   :emphasize-lines: 1, 8
+
+   from clastic import POST
+
+
+   def create_app():
+       static_app = StaticApplication(STATIC_PATH)
+       routes = [
+           ("/", home, "home.html"),
+           POST("/submit", add_entry, render_add_entry),
+           ("/static", static_app),
+       ]
+
+       ...
+
+
+We add this route as a :func:`POST <clastic.POST>` route.
+This makes sure that other HTTP methods will not be allowed for this path.
+You can try typing the address ``http://localhost:5000/submit``
+into the location bar of your browser,
+and you should see a "method not allowed" error.
+There are also other method-restricted routes,
+like :func:`GET <clastic.GET>`, :func:`PUT <clastic.PUT>`, and
+:func:`DELETE <clastic.DELETE>`.
+
+
+Named path segments
+-------------------
+
+Now let's turn to using the stored shortened links.
+When the shortened link URL is visited,
+we just want to redirect the browser to the target URL.
+This is going to be a GET-only route:
+
+.. code-block:: python
+   :emphasize-lines: 8
+
+   from clastic import GET
+
+
+   routes = [
+       ("/", home, "home.html"),
+       POST("/submit", add_entry, render_add_entry, middlewares=[new_link_mw]),
+       ("/static", static_app),
+       GET('/<alias>', use_entry),
+   ]
+
+
+The "alias" part of the path is placed between angular brackets
+to make that segment a named parameter to the endpoint function.
+And there is no need for a renderer in this case;
+the endpoint function will do the redirection right away:
+
+.. code-block:: python
+
+   def use_entry(alias, db):
+       entry = db.get_link(alias)
+       return redirect(entry["target"], code=HTTPStatus.MOVED_PERMANENTLY)
+
+
+Using middleware
+----------------
+
+Clastic allows us to use :doc:`middleware <middleware>`
+to keep application logic out of endpoint functions and to promote reuse.
+For example,
+the :class:`PostDataMiddleware <clastic.middleware.form.PostDataMiddleware>`
+can be used to convert the form data into appropriate types
+and make them available to endpoint functions as parameters:
+
+.. code-block:: python
+
+   from clastic.middleware.form import PostDataMiddleware
+
+
+   def create_app():
+       new_link_mw = PostDataMiddleware(
+           {"target_url": str, "alias": str, "max_count": int, "expiry_time": str}
+       )
+
+       static_app = StaticApplication(STATIC_PATH)
+       routes = [
+           ("/", home, "home.html"),
+           POST("/submit", add_entry, render_add_entry, middlewares=[new_link_mw]),
+           ("/static", static_app),
+       ]
+
+       ...
+
+
+The endpoint function doesn't need to get the data from ``request.values``
+anymore:
+
+.. code-block:: python
+
+   def add_entry(db, target_url, alias, expiry_time, max_count):
+       entry = db.add_link(
+           target_url=target_url, alias=alias, expiry_time=expiry_time, max_count=max_count
+       )
+       return {}
```

Comment:
At that stage in the tutorial, the render function wasn't using the context, so I had left it to some later stage. But I've rewritten that part since.